cmd/otel-allocator/allocation/allocator.go

```diff
@@ -88,119 +88,129 @@ func (allocator *Allocator) Collectors() map[string]*collector {
 	return collectorsCopy
 }
 
+// findNextCollector finds the next collector with fewer number of targets.
+// This method is called from within SetWaitingTargets and SetCollectors, whose caller
+// acquires the needed lock.
+func (allocator *Allocator) findNextCollector() *collector {
+	var col *collector
+	for _, v := range allocator.collectors {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col == nil {
+			col = v
+		} else {
+			if v.NumTargets < col.NumTargets {
+				col = v
+			}
+		}
+
+	}
+	return col
+}
+
+// assignTargetToNextCollector assigns a target to the next available collector
+func (allocator *Allocator) assignTargetToNextCollector(target *TargetItem) {
+	chosenCollector := allocator.findNextCollector()
+	targetItem := TargetItem{
+		JobName:   target.JobName,
+		Link:      LinkJSON{fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL: target.TargetURL,
+		Label:     target.Label,
+		Collector: chosenCollector,
+	}
+	allocator.targetItems[targetItem.hash()] = &targetItem
+	chosenCollector.NumTargets++
+	targetsPerCollector.WithLabelValues(chosenCollector.Name).Set(float64(chosenCollector.NumTargets))
```

Comment:
```suggestion
        if _, found := allocator.targetItems[target.hash()]; found {
		target.Collector.NumTargets--
		targetsPerCollector.WithLabelValues(target.Collector.Name).Set(float64(target.Collector.NumTargets))
	}
	allocator.targetItems[targetItem.hash()] = &targetItem
	chosenCollector.NumTargets++
	targetsPerCollector.WithLabelValues(chosenCollector.Name).Set(float64(chosenCollector.NumTargets))
```
We have to update the metrics if the target has already been assigned

Reply:
This method assumes that the target hasn't been assigned already. This is only called after the collectors are cleared or when a new target has been found in the `tempTargetMap`. If we were to decrement the number of targets here we would be reporting incorrectly

Reply:
This will be the case if you apply my suggestion.    
If you don't, then this assumption should be part of the Method documentation.

Reply:
this is only the case where targets are reassigned i believe, maybe the right solution is to just set the metric to 0 when we remove the unused collectors.

Reply:
i will also include that assumption in the comment.

Reply:
honestly, the right solution here would be an asynchronous gauge instrument that tracks this information so we don't need to be smart about updating it. cc @Aneurysm9 for when we migrate this to the otel sdk

Reply:
If we keep this assumption we have to set it to  0 and let it then disappear.  Else the aggregations will show false data.

Reply:
Yep, i set it to 0 now. Because of how prometheus works, you can't delete the information entirely though. Does the approach I have out there now look correct?